src/librustc_middle/mir/interpret/error.rs

```diff
@@ -90,7 +81,7 @@ impl<'tcx> ConstEvalErr<'tcx> {
 
     pub fn report_as_error(&self, tcx: TyCtxtAt<'tcx>, message: &str) -> ErrorHandled {
         match self.struct_error(tcx, message, |mut e| e.emit()) {
-            Ok(_) => ErrorHandled::Reported,
+            Ok(_) => ErrorHandled::Reported(ErrorReported),
```

Comment:
Is still seems rather easy to just say `ErrorReported` when the compiler says that that is the type you need. Maybe the only way to construct this should be something like `ErrorReported::assert_error_reported_to_user` or so? Could that constructor do `delay_bug` just to be extra safe or does that not make sense?

Reply:
@RalfJung We're working on that but it's not ready yet because of some abuse in the compiler (of `ty::Error` which we wanted to add an `ErrorReported` to).

I tried to make this PR not blocked on that, but write `ErrorReported` or `error_reported` everywhere relevant so e.g. @mark-i-m can find it later.

This is mostly because I only did this PR as a prerequisite for #70970.
I am fine with blocking this PR on @mark-i-m's work if fixing #70942 is not a priority.

Reply:
@eddyb sure, if improving this is on your list then that's fine for me. I was just pointing out that as the final solution this still seems suboptimal.

Reply:
> I was just pointing out that as the final solution this still seems suboptimal.

Indeed. Also, I should've pointed at #69426.

Reply:
Having this PR will definitely make my next PR easier.